modules/sfp_spyonweb.py

```diff
@@ -0,0 +1,367 @@
+        'api_key': '',
+        'limit': 100,
+        'timeout': 30,
+        'maxage': 1095,   # 3 years
+        'verify': True,
+        'cohostsamedomain': False,
+        'maxcohost': 100
+    }
+
+    # Option descriptions
+    optdescs = {
+        'api_key': "SpyOnWeb API key.",
+        'limit': "Maximum number of results to fetch.",
+        'timeout': "Query timeout, in seconds.",
+        'maxage': "The maximum age of the data returned, in days, in order to be considered valid.",
+        'verify': "Verify co-hosts are valid by checking if they still resolve to the shared IP.",
+        'cohostsamedomain': "Treat co-hosted sites on the same target domain as co-hosting?",
+        'maxcohost': "Stop reporting co-hosted sites after this many are found, as it would likely indicate web hosting.",
+    }
+
+    cohostcount = 0
+    results = None
+    errorState = False
+
+    def setup(self, sfc, userOpts=dict()):
+        self.sf = sfc
+        self.__dataSource__ = "SpyOnWeb"
+        self.results = dict()
+        self.cohostcount = 0                                                                                                                                                                                       
+
+        for opt in userOpts.keys():
+            self.opts[opt] = userOpts[opt]
+
+    # Verify a host resolves to an IP
+    def validateIP(self, host, ip):
+        try:
+            addrs = socket.gethostbyname_ex(host)
+        except BaseException as e:
+            self.sf.debug("Unable to resolve " + host + ": " + str(e))
+            return False
+
+        for addr in addrs:
+            if type(addr) == list:
+                for a in addr:
+                    if str(a) == ip:
+                        return True
+            else:
+                if str(addr) == ip:
+                    return True
+        return False
+
+    # What events is this module interested in for input
+    def watchedEvents(self):
+        return ['IP_ADDRESS', 'INTERNET_NAME', 'DOMAIN_NAME', 'WEB_ANALYTICS_ID']
+
+    # What events this module produces
+    def producedEvents(self):
+        return ['CO_HOSTED_SITE', 'INTERNET_NAME', 'AFFILIATE_DOMAIN', 'WEB_ANALYTICS_ID']
+
+    # Retrieve hosts with the specified Google Analytics ID
+    # https://api.spyonweb.com/v1/docs
+    def queryGoogleAnalytics(self, qry, limit=100):
+        url = "https://api.spyonweb.com/v1/analytics/" + qry
+        url += "?limit=" + str(limit)
+        url += "&access_token=" + self.opts['api_key']
+
+        res = self.sf.fetchUrl(url, timeout=self.opts['timeout'], useragent=self.opts['_useragent'])
+
+        if res['content'] is None:
+            self.sf.info("No results found for " + qry)
+            return None
+
+        try:
+            data = json.loads(res['content'])
+        except Exception as e:
+            self.sf.debug("Error processing JSON response.")
+            return None
+
+        if not data['status'] == 'found':
```

Comment:
Please use .get() in all the cases you are referring to dictionary keys, as it will possibly raise exceptions when the 'status' key isn't set. Also in other cases the key may be set but may be zero, which will result in a different exception.

Reply:
I've changed to `get()` in the instances where I think this is reasonable.

Reply:
The other should still have it. For instance, below on 108 you correctly check if result is a key in data, but then further down on 112 you are assuming that data['result'] is not going to be None. If it is though, an exception will be raised and the whole scan will fail. Same for various others below.